Add get_stat() API to parse /proc/stat

kfiros opened the pull request.

dtrugman commented on src/procfs.cpp:
```diff
@@ -137,6 +140,50 @@ uptime procfs::get_uptime() const
     return parsers::parse_uptime_line(line);
 }
 
+sys_stat procfs::get_stat() const
```
So there was actually already a parser for files like that.
It's called `file_parser` and it's defined in `parsers.hpp`.
It goes over the file line by line, splits the line into two and then calls the right parser from an unordered map of parsers.
This is the way we parse the `/proc/<pid>/status` file (The `status.cpp` file is a great reference).
The only thing that you should probably change is to get this class accept the DELIM as a parameter instead of using the hard coded ':'.

dtrugman replied:
I guess only this method should be replaced with the parser instance. The per line parsers are still going to be required obviously.

kfiros replied:
You are right, except the number of CPUs is not constant : cpu1 cpu2 ... So using a map might be problematic here, unless I add some specific changes (for example if we have cpuX, remove the X and find cpu). Maybe we can solve that by getting a function as a parameter which manipulates the keys and only then calls the suitable function from the map. What do you think?

dtrugman replied:
Hmm, that's a good point...
I like your idea. I think that we can pass that key "permutation" function when constructing the parser.
There are a few things to consider though:
1. The function should probably return a copy of the key? For example, return `cpuX` from `cpu1` because we still need to pass the actual key to the line parsers. I guess we'll need the value `1` to add that to the correct index in the cpus `vector`. I think that we cannot assume correct alphanumeric order with more than 10 cpus (possibly ordered: 1, 10, 11, ..., 2, 3, ....)
2. Today we don't even pass the key to the parser, this should probably change as well?
3. On the other hand, we don't want to copy all the keys when there's no good reason

I guess this is going to be challenging. Best way is probably to try and see how it turns out, but I believe you can find the right solution :)
If it makes things too complicated, we can re-consider using the current solution, so make sure you stash it somewhere.

kfiros replied:
1. As I see it - we need another function to provide another level of translation. For example in our case:
cpuX -> cpu

So all the cpu lines will be parsed in the same manner.

2. I don't think we need to care about the order of the CPUs, we should add them in the same order as the lines order.

dtrugman commented on include/pfs/parsers.hpp:
```diff
@@ -63,12 +63,16 @@ void parse_lines(
 zone parse_buddyinfo_line(const std::string& line);
 load_average parse_loadavg_line(const std::string& line);
 uptime parse_uptime_line(const std::string& line);
+sys_stat::cpu_info parse_stat_cpu_line(const std::string& line);
```
Since we don't expect anyone to call this parser directly, we always declare per-line parsers as static methods inside the relevant cpp file.

dtrugman commented on include/pfs/parsers.hpp:
```diff
@@ -63,12 +63,16 @@ void parse_lines(
     }
 }
 
+std::string parse_key_value_line(const std::string& line);
```
See my comment about the `file_parser` class

dtrugman commented on src/procfs.cpp:
```diff
@@ -137,6 +140,50 @@ uptime procfs::get_uptime() const
+
+    for (const std::string& line : lines) {
+        if (line.rfind("cpu", 0) == 0) {
+            stats.cpus.emplace_back(parsers::parse_stat_cpu_line(line));
```
Following your suggestion and my comment about the parser, I'm not sure this is a correct assumption. I mean, not sure CPUs are going to be always ordered correctly.
Especially with hot-swapping and such, I wonder how this is going to be handled by the machine...
Maybe we should use the `X` from `cpuX` to add the values to the right index in the vector...

dtrugman commented on src/procfs.cpp:
```diff
@@ -137,6 +140,50 @@ uptime procfs::get_uptime() const
+
+    for (const std::string& line : lines) {
+        if (line.rfind("cpu", 0) == 0) {
+            stats.cpus.emplace_back(parsers::parse_stat_cpu_line(line));
```
Following your suggestion and my comment about the parser, I'm not sure this is a correct assumption. I mean, not sure CPUs are going to be always ordered correctly.
Especially with hot-swapping and such, I wonder how this is going to be handled by the machine...
Maybe we should use the `X` from `cpuX` to add the values to the right index in the vector...

kfiros replied:
I walked over some recent kernel code, and it seems ordered, can you verify me?

https://github.com/torvalds/linux/blob/5bfc75d92efd494db37f5c4c173d3639d4772966/fs/proc/stat.c#L165

```
for_each_online_cpu() ->
#define for_each_online_cpu(cpu)   for_each_cpu((cpu), cpu_online_mask)

```
```
/**
 * for_each_cpu - iterate over every cpu in a mask
 * @cpu: the (optionally unsigned) integer iterator
 * @mask: the cpumask pointer
 *
 * After the loop, cpu is >= nr_cpu_ids.
 */
#define for_each_cpu(cpu, mask)				\
	for ((cpu) = -1;				\
		(cpu) = cpumask_next((cpu), (mask)),	\
		(cpu) < nr_cpu_ids;)


```
```

/**
 * cpumask_next - get the next cpu in a cpumask
 * @n: the cpu prior to the place to search (ie. return will be > @n)
 * @srcp: the cpumask pointer
 *
 * Returns >= nr_cpu_ids if no further cpus set.
 */
unsigned int cpumask_next(int n, const struct cpumask *srcp)
{
	/* -1 is a legal arg here. */
	if (n != -1)
		cpumask_check(n);
	return find_next_bit(cpumask_bits(srcp), nr_cpumask_bits, n + 1);
}
EXPORT_SYMBOL(cpumask_next);
```

dtrugman replied:
Looks good. I think we can assume they are ordered till proven otherwise :)

dtrugman commented on sample/format.hpp:
```diff
@@ -587,6 +588,69 @@ inline std::ostream& operator<<(std::ostream& out,
+    out << "stat_softirq[";
+    out << "total: " << stats.softirqs.total << ", ";
+    out << "per_softirq: ";
+    for (auto it = stats.softirqs.per_item.begin();
```
Same... use `join`?

kfiros replied:
Done

dtrugman commented on src/parsers/proc_stat.cpp:
```diff
@@ -0,0 +1,219 @@
+        utils::stot(tokens[NICE], out.nice);
+        utils::stot(tokens[SYSTEM], out.system);
+        utils::stot(tokens[IDLE], out.idle);
+        if (tokens.size() > IOWAIT) {
```
Can you please add newlines around `if` blocks? (these are not `if .. else`-s)

dtrugman commented on src/parsers/proc_stat.cpp:
```diff
@@ -0,0 +1,219 @@
+static void parse_cpu(const std::string& value, proc_stat& out)
+{
+    proc_stat::cpu cpu;
+    memset(&cpu, 0, sizeof(cpu));
```
A better implementation will assign default values to fields in the `cpu` struct

kfiros commented on include/pfs/parsers.hpp:
```diff
@@ -80,13 +81,19 @@ netlink_socket parse_netlink_socket_line(const std::string& line);
                  const char delim,
-                 const std::set<std::string>& keys = {})
+                 const std::set<std::string>& keys = {},
+                 RemapFunction& remap = nullptr)
```
I called it key_remap, I think it's redundant to write the word function in the name, because the type already informs you about that, anyway your call.

dtrugman replied:
`key_remap` is great!

kfiros commented on include/pfs/parsers.hpp:
```diff
@@ -80,13 +81,19 @@ netlink_socket parse_netlink_socket_line(const std::string& line);
                  const char delim,
-                 const std::set<std::string>& keys = {})
+                 const std::set<std::string>& keys = {},
+                 RemapFunction& remap = nullptr)
```
Although your intention is right, I prefer the user won't need to set the remap (even to nullptr). In the current implementation, the user can just:
parse(path, delim, keys); or parse(path, delim);

If we will switch, and the user provides a set of keys:
parse(path, delim, nullptr, keys);

Fix me if I'm wrong.

dtrugman replied:
I just looked at the code again and realized something.
I think that the `delim` and `remap` arguments should be passed to the `file_parser` during construction.
These are constant per parser and I don't see a reason to have the same parsers call the same method with different `delim` and `remap` arguments.... wdyt?

kfiros commented on include/pfs/types.hpp:
```diff
@@ -359,6 +360,38 @@ struct load_average
+        std::vector<unsigned long long> per_item;
+    };
+
+    std::vector<cpu> cpus;
```
I'm not sure... In the CPUs it is very easy to use from the user perspective - for example, it is very convenient for the user to get the idle time... 

In the interrupts/softirqs, the access is by the IRQ ID. For example, from the man:
 each  subsequent  column is the total for that particular numbered interrupt.

dtrugman replied:
I agree that it is easy to get the idle time for a CPU, but to me it's very confusing to find the total stats for all the cpus in cpus[0], stats for CPU0 in cpus[1], CPU1 in cpus[2], etc...

kfiros commented on sample/format.hpp:
```diff
@@ -587,6 +588,69 @@ inline std::ostream& operator<<(std::ostream& out,
+                                const pfs::proc_stat& stats)
+{
+    size_t cpu_idx = 0;
+    for (auto cpu = stats.cpus.begin(); cpu != stats.cpus.end(); cpu++) {
```
Changed it to a regular for, I want the index, to make the printing right..

kfiros commented on sample/format.hpp:
```diff
@@ -587,6 +588,69 @@ inline std::ostream& operator<<(std::ostream& out,
+    out << "stat_interrupts[";
+    out << "total: " << stats.interrupts.total << ", ";
+    out << "per_interrupt: ";
+    for (auto it = stats.interrupts.per_item.begin();
```
I decided no to join the cpu, as I want to display each field title (user: x, idle: y)...

kfiros commented on include/pfs/types.hpp:
```diff
@@ -359,6 +360,38 @@ struct load_average
     int last_created_task;
 };
 
+struct proc_stat
```
Used the original names.

dtrugman commented on include/pfs/types.hpp:
```diff
@@ -359,6 +360,38 @@ struct load_average
+    struct sequence
+    {
+        unsigned long long total;
+        std::vector<unsigned long long> per_item;
```
Good point! I'm trying to think if I have a good idea here... For now, let's keep it...

kfiros commented:
Hi, I made some changes... I think it's much better now.

Now there is a template sequence struct, which can fit the CPUs and other types as well.
So now - we have sequence<cpu> cpus; and cpus.total is the first line, but cpus.per_item[0] is CPU0.

I also updated the file parsers, so it will get the delimiter and remap_function.

Another important thing - I changed the parser function to get 3 arguments, including the original key. Now I do test for the original key in case of parsing the CPUs.

Finally, I ran the clang format.

dtrugman commented on include/pfs/parsers.hpp:
```diff
@@ -101,28 +106,36 @@ class file_parser
             std::tie(key, value) = utils::split_once(line, _delim);
-            if (key.empty())
+
+            std::string remapped_key = key;
```
Maybe this part should come after the `utils::rtrim`... this way we don't have to do it twice.

dtrugman commented on include/pfs/parsers.hpp:
```diff
@@ -101,28 +106,36 @@ class file_parser
                 utils::ltrim(value);
                 auto& parser = iter->second;
-                parser(value, output);
+                parser(key, value, output);
```
Sorry for throwing another idea here, but I think we can improve further...
Consider the following idea. If the key remap was the following one:
```
void cpu_remap(std::string& key)
{
    static const std::string CPU("cpu");
    if (key.size() == CPU.size() + 1 &&
        key.compare(CPU, 0, CPU.size()) == 0 &&)
    {
        key[CPU.size()] = 'X';
    }
}
```
That way, all specific CPU entries like "cpu1", "cpu2", etc are mapped to "cpuX" for which you can have a separate handler from the total "cpu" entry.
This way you don't need to pass key, and thus avoiding the extra copy per key. WDYT?

kfiros replied:
I think it's OK to pass the key, that way we can give flexibility for other cases in the future, don't you think it's better than just handling this specific case?

dtrugman replied:
I agree that this gives flexibility, but on the other hand this copies every single key.
That is fine if there's an actual use case, but at the moment we lose performance and simplicity (all the callbacks receive an unused extra parameter) over nothing...
I think we should treat it as YAGNI till we do. Hope you don't object strongly to my opinion here.

dtrugman commented on test/proc_stat.cpp:
```diff
@@ -0,0 +1,36 @@
+
+TEST_CASE("Parse stat", "[procfs][proc_stat]")
+{
+    static constexpr auto Deviation     = 30;
```
Mind using `DEVIATION` and `CPU_MIN_ENTRIES`?

kfiros commented:
1. Fixed the remap function - now it gets a reference and doesn't return a value. It would save the extra copy.
2. Fixed the /proc/stat remap function, now we check if the key is "cpu" / begins with "cpu" / other.
